diff --git a/node_modules/nx/src/daemon/server/plugins.js b/node_modules/nx/src/daemon/server/plugins.js
index b30628a..1ceaf6f 100644
--- a/node_modules/nx/src/daemon/server/plugins.js
+++ b/node_modules/nx/src/daemon/server/plugins.js
@@ -13,7 +13,7 @@ async function getPlugins() {
     const pluginsConfiguration = (0, nx_json_1.readNxJson)().plugins ?? [];
     const [result, cleanupFn] = await (0, internal_api_1.loadNxPlugins)(pluginsConfiguration, workspace_root_1.workspaceRoot);
     cleanup = cleanupFn;
-    return result;
+    return loadedPlugins = result;
 }
 exports.getPlugins = getPlugins;
 function cleanupPlugins() {
